salt/_auth/kubernetes_rbac.py

```diff
@@ -1,31 +1,152 @@
-            token == ADMIN_TOKEN,
-        ]),
-        'groups': lambda _u, _t: [
-            'node_admin',
-        ],
-    },
+    if MISSING_DEPS:
+        return False, 'Missing Python dependencies: {}'.format(
+            ', '.join(MISSING_DEPS))
+    else:
+        return __virtualname__
+
+
+AUTH_HANDLERS = {}
+
+
+def _log_exceptions(f):
+    def wrapped(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except:
+            log.exception('Exception thrown')
+            raise
+
+    return wrapped
+
+
+@_log_exceptions
+def _auth_basic(kubeconfig, username, token):
+    decoded = base64.decodestring(token)
+    if ':' not in decoded:
+        log.warning('Invalid Basic token format: missing ":"')
+        return False
+
+    (token_username, _) = decoded.split(':', 1)
+
+    if token_username != username:
+        log.warning('Invalid Basic token: username mismatch')
+        return False
+
+    # Using the '/version/' endpoint which is unauthenticated by default but,
+    # when presented authentication data, will process this information and fail
+    # accordingly.
+    url = '{}/version/'.format(kubeconfig.host)
+    verify = kubeconfig.ssl_ca_cert if kubeconfig.verify_ssl else False
+    try:
+        response = requests.get(
+            url,
+            headers={
+                'Authorization': 'Basic {}'.format(token),
+            },
+            verify=verify,
+        )
+
+        if 200 <= response.status_code < 300:
+            return True
+        else:
+            return False
+    except:
+        log.exception('Error during request')
+        raise
+
+    return False
+
+
+@_log_exceptions
+def _groups_basic(kubeconfig, username, token):
+    kubeconfig.api_key = {
+        'authorization': token,
+    }
+    kubeconfig.api_key_prefix = {
+        'authorization': 'Basic',
+    }
+    kubeconfig.username = username
+    kubeconfig.password = None
+    kubeconfig.cert_file = None
+    kubeconfig.key_file = None
+
+    client = kubernetes.client.ApiClient(configuration=kubeconfig)
+
+    authz_api = kubernetes.client.AuthorizationV1Api(api_client=client)
+
+    groups = set()
+
+    result = authz_api.create_self_subject_access_review(
+        body=kubernetes.client.V1SelfSubjectAccessReview(
+            spec=kubernetes.client.V1SelfSubjectAccessReviewSpec(
+                resource_attributes=kubernetes.client.V1ResourceAttributes(
+                    resource='nodes',
+                    verb='*',
+                ),
+            ),
+        ),
+    )
+
+    if result.status.allowed:
+        groups.add('node-admins')
+
+    return list(groups)
+
+
+AUTH_HANDLERS['basic'] = {
+    'auth': _auth_basic,
+    'groups': _groups_basic,
 }
 
 
+@_log_exceptions
+def _load_kubeconfig(opts):
+    config = {
+        'kubeconfig': None,
+        'context': None,
+    }
+
+    for opt in opts['external_auth'][__virtualname__]:
+        if opt.startswith('^'):
```

Comment:
Would you mind adding a comment for this "trick"? I was confused on first read

Reply:
It's (apparently, though admittedly not very fond of it) 'standard' in Salt `eauth` modules and how to configure them, e.g. the built-in `file` and `ldap` ones (which is where I got this idea).

Reply:
Oh, wasn't aware of it... Documentation of our internal modules should be improved in the future, so re-using/maintaining them is easier even for non-experts.